diff --git a/entity_reference_edit_link.module b/entity_reference_edit_link.module
index 9d8b89b..60eee19 100644
--- a/entity_reference_edit_link.module
+++ b/entity_reference_edit_link.module
@@ -190,7 +190,7 @@ function _entity_reference_edit_link_build_entity_type_link($node) {
  * Implements custom function to check allowed content types.
  */
 function _entity_reference_edit_link_allowed_content_type($node) {
-  return !empty($allowedContentTypes = \Drupal::config('edit_link.config')->get('content_types'))
+  return !empty($allowedContentTypes = \Drupal::config('entity_reference_edit_link.config')->get('content_types'))
     && in_array($node->bundle(), $allowedContentTypes);
 }
 
diff --git a/src/Form/EditLinkConfigForm.php b/src/Form/EditLinkConfigForm.php
index 64c4c2e..e1a4a84 100644
--- a/src/Form/EditLinkConfigForm.php
+++ b/src/Form/EditLinkConfigForm.php
@@ -16,7 +16,7 @@ class EditLinkConfigForm extends ConfigFormBase {
    *
    * @const string
    */
-  protected const SETTINGS = 'edit_link.config';
+  protected const SETTINGS = 'entity_reference_edit_link.config';
 
   /**
    * The entity type manager.
