diff --git a/core/modules/content_moderation/src/Plugin/views/ModerationStateJoinViewsHandlerTrait.php b/core/modules/content_moderation/src/Plugin/views/ModerationStateJoinViewsHandlerTrait.php
index 3affab6..ee28090 100644
--- a/core/modules/content_moderation/src/Plugin/views/ModerationStateJoinViewsHandlerTrait.php
+++ b/core/modules/content_moderation/src/Plugin/views/ModerationStateJoinViewsHandlerTrait.php
@@ -35,6 +35,15 @@ trait ModerationStateJoinViewsHandlerTrait {
             'field' => 'content_entity_type_id',
             'value' => $left_entity_type->id(),
           ],
+          [
+            'field' => 'content_entity_id',
+            'left_table' => $table_alias,
+            'left_field' => $left_entity_type->getKey('id'),
+          ],
+          [
+            'field' => 'workflow',
+            'value' => $this->getWorkflowIds(),
+          ],
         ],
       ];
       if ($left_entity_type->isTranslatable()) {
