diff --git a/js/csv-field.js b/js/csv-field.js
index 5f9c3f3ac61a5e969be2f9e5639a130cf2a2ead6..597beb061db276a9ddcd250cd8f3406376725170 100644
--- a/js/csv-field.js
+++ b/js/csv-field.js
@@ -127,9 +127,7 @@
                 dtable.search("").draw();

                 var $dtLayoutElements = dtable.tables().containers().to$()
-                  .find("div.dt-layout-table")
-                  .add("div.dt-info")
-                  .add("div.dt-paging");
+                  .find("div.dt-layout-table, div.dt-info, div.dt-paging");
                 $dtLayoutElements.hide();

                 var $search = $(table).parent().find('input[type="search"]');
