diff --git a/core/modules/jsonapi/src/Serializer/Serializer.php b/core/modules/jsonapi/src/Serializer/Serializer.php
index 93669b50402..2db2159fc7c 100644
--- a/core/modules/jsonapi/src/Serializer/Serializer.php
+++ b/core/modules/jsonapi/src/Serializer/Serializer.php
@@ -22,7 +22,7 @@
  * @see https://www.drupal.org/project/drupal/issues/3032787
  * @see jsonapi.api.php
  */
-final class Serializer extends SymfonySerializer implements JsonSchemaProviderSerializerInterface {
+class Serializer extends SymfonySerializer implements JsonSchemaProviderSerializerInterface {
 
   use JsonSchemaProviderSerializerTrait;
 
