diff --git a/modules/jsonapi/src/Serializer/Serializer.php b/modules/jsonapi/src/Serializer/Serializer.php
index 0ea0442..bf2af7f 100644
--- a/modules/jsonapi/src/Serializer/Serializer.php
+++ b/modules/jsonapi/src/Serializer/Serializer.php
@@ -16,7 +16,7 @@ use Symfony\Component\Serializer\Serializer as SymfonySerializer;
  *
  * @internal
  */
-final class Serializer extends SymfonySerializer {
+class Serializer extends SymfonySerializer {
 
   /**
    * A normalizer to fall back on when JSON:API cannot normalize an object.
