diff --git a/core/misc/tabledrag.js b/core/misc/tabledrag.js
index bd25008347..49092f062c 100644
--- a/core/misc/tabledrag.js
+++ b/core/misc/tabledrag.js
@@ -66,6 +66,10 @@
     const self = this;
     const $table = $(table);

+    // Store table drag object settings to allow
+    // async row insertions and using makeDraggable.
+    $table.data('tableDragObject', this);
+
     /**
      * @type {jQuery}
      */
