diff --git a/js/csv-field.js b/js/csv-field.js
index a14e262..a4cca8c 100644
--- a/js/csv-field.js
+++ b/js/csv-field.js
@@ -71,7 +71,7 @@
                     // by default collapse details row
                     settings.columns = [
                       {
-                        className: "dtr-control",
+                        className: "dtr-control dtr-hidden",
                         orderable: false,
                         data: null,
                         title: "",
@@ -97,7 +97,18 @@
                       ...settings,
                       search: {
                         return: true
-                      }
+                      },
+                      headerCallback: function(thead, data, start, end, display) {
+                        var el = $(thead).find('th.dtr-hidden');
+                        $(el).addClass('ma__visually-hidden');
+                        $(el).css('max-width', '4px');
+                        $(el).css('position', 'initial');
+                        $(el).css('white-space', 'nowrap');
+                        $(el).css('pointer-events', 'none');
+                        $(el).css('color', 'transparent');
+                        $(el).removeAttr('aria-label');
+                        $(el).text('a column of expand buttons (see additional information of each row)');
+                      },
                     });
 
                     // Clear search input and redraw table
@@ -117,7 +128,7 @@
                     $(table).siblings('.dataTables_info').hide();
                     $(table).css('border-bottom', 'unset');
                     var $search = $(table).parent().find('input[type="search"]');
-                    var $button = $('<button type="submit" class="dt-search-submit" value="Submit">Submit</button>').insertAfter($search);
+                    var $button = $('<button type="submit" class="dt-search-submit ma__button   ma__button--secondary  ma__button--lowercase" value="Submit">Submit</button>').insertAfter($search);
 
                     // Event listener to show rows only when a search is performed
                     dtable.on('search.dt', function(e) {
@@ -145,11 +156,37 @@
                     });
                   }
                   else {
-                    $(table).DataTable(settings);
+                    $(table).DataTable({
+                      ...settings,
+                      headerCallback: function(thead, data, start, end, display) {
+                        let el = $(thead).find('th.dtr-hidden');
+                        $(el).addClass('ma__visually-hidden');
+                        $(el).css('max-width', '4px');
+                        $(el).css('position', 'initial');
+                        $(el).css('white-space', 'nowrap');
+                        $(el).css('pointer-events', 'none');
+                        $(el).css('color', 'transparent');
+                        $(el).removeAttr('aria-label');
+                        $(el).text('a column of expand buttons (see additional information of each row)');
+                      }
+                    });
                   }
                 }
                 else {
-                  $(table).DataTable(settings);
+                  $(table).DataTable({
+                    ...settings,
+                    headerCallback: function(thead, data, start, end, display) {
+                      let el = $(thead).find('th.dtr-hidden');
+                      $(el).addClass('ma__visually-hidden');
+                      $(el).css('max-width', '4px');
+                      $(el).css('position', 'initial');
+                      $(el).css('white-space', 'nowrap');
+                      $(el).css('pointer-events', 'none');
+                      $(el).css('color', 'transparent');
+                      $(el).removeAttr('aria-label');
+                      $(el).text('a column of expand buttons (see additional information of each row)');
+                    }
+                  });
                 }
               } else {
                 var lines = parsed.data,
