diff --git a/js/csv-field.js b/js/csv-field.js
index 516fbef..74ee315 100644
--- a/js/csv-field.js
+++ b/js/csv-field.js
@@ -73,7 +73,7 @@
                     // by default collapse details row
                     settings.columns = [
                       {
-                        className: "dtr-control",
+                        className: "dtr-control dtr-hidden",
                         orderable: false,
                         data: null,
                         title: "",
@@ -128,6 +128,17 @@
                       language: {
                         search: settings.searchLabel ? settings.searchLabel + ':' : 'Search:'
                       },
+                      headerCallback: function(thead, data, start, end, display) {
+                        var el = $(thead).find('th.dtr-hidden');
+                        $(el).addClass('ma__visually-hidden');
+                        $(el).css('max-width', '4px');
+                        $(el).css('position', 'initial');
+                        $(el).css('white-space', 'nowrap');
+                        $(el).css('pointer-events', 'none');
+                        $(el).css('color', 'transparent');
+                        $(el).removeAttr('aria-label');
+                        $(el).text('a column of expand buttons (see additional information of each row)');
+                      },
                     });

                     // Clear search input and redraw table
@@ -140,7 +151,7 @@
                     $(table).siblings('.dataTables_info').hide();
                     $(table).css('border-bottom', 'unset');
                     var $search = $(table).parent().find('input[type="search"]');
-                    var $button = $('<button type="submit" class="dt-search-submit" value="Search">Search</button>').insertAfter($search);
+                    var $button = $('<button type="submit" class="dt-search-submit ma__button ma__button--secondary ma__button--lowercase" value="Search">Search</button>').insertAfter($search);

                     // Event listener to show rows only when a search is performed
                     dtable.on('search.dt', function(e) {
@@ -168,11 +179,37 @@
                     });
                   }
                   else {
-                    $(table).DataTable(settings);
+                    $(table).DataTable({
+                      ...settings,
+                      headerCallback: function(thead, data, start, end, display) {
+                        var el = $(thead).find('th.dtr-hidden');
+                        $(el).addClass('ma__visually-hidden');
+                        $(el).css('max-width', '4px');
+                        $(el).css('position', 'initial');
+                        $(el).css('white-space', 'nowrap');
+                        $(el).css('pointer-events', 'none');
+                        $(el).css('color', 'transparent');
+                        $(el).removeAttr('aria-label');
+                        $(el).text('a column of expand buttons (see additional information of each row)');
+                        },
+                    });
                   }
                 }
                 else {
-                  $(table).DataTable(settings);
+                  $(table).DataTable({
+                    ...settings,
+                    headerCallback: function(thead, data, start, end, display) {
+                      var el = $(thead).find('th.dtr-hidden');
+                      $(el).addClass('ma__visually-hidden');
+                      $(el).css('max-width', '4px');
+                      $(el).css('position', 'initial');
+                      $(el).css('white-space', 'nowrap');
+                      $(el).css('pointer-events', 'none');
+                      $(el).css('color', 'transparent');
+                      $(el).removeAttr('aria-label');
+                      $(el).text('a column of expand buttons (see additional information of each row)');
+                    },
+                  });
                 }
                 var className = 'dt-search-right-aligned';
                 if (typeof settings.lengthChange != "undefined" && typeof settings.searching != "undefined") {
