diff --git a/js/csv-field.js b/js/csv-field.js
index c235d0f43ea1f24fb34cd3fcd98836234840a24c..5f9c3f3ac61a5e969be2f9e5639a130cf2a2ead6 100644
--- a/js/csv-field.js
+++ b/js/csv-field.js
@@ -118,14 +118,19 @@
                   language: {
                     search: tableSettings.searchLabel ? tableSettings.searchLabel + ":" : "Search:",
                   },
+                  layout: {
+                    topStart: "search",
+                    topEnd: ""
+                  }
                 });

                 dtable.search("").draw();
-                dtable.rows().nodes().to$().hide();
-                $(table).find("thead").hide();
-                $(table).siblings(".dataTables_paginate").hide();
-                $(table).siblings(".dataTables_info").hide();
-                $(table).css("border-bottom", "unset");
+
+                var $dtLayoutElements = dtable.tables().containers().to$()
+                  .find("div.dt-layout-table")
+                  .add("div.dt-info")
+                  .add("div.dt-paging");
+                $dtLayoutElements.hide();

                 var $search = $(table).parent().find('input[type="search"]');
                 var $button = $('<button type="submit" class="dt-search-submit">Search</button>')
@@ -134,17 +139,9 @@
                 dtable.on("search.dt", function () {
                   var searchTerm = dtable.search();
                   if (searchTerm) {
-                    dtable.rows().nodes().to$().show();
-                    $(table).find("thead").show();
-                    $(table).css("border-bottom", "1px solid rgba(0, 0, 0, 0.3)");
-                    $(table).siblings(".dataTables_paginate").show();
-                    $(table).siblings(".dataTables_info").show();
+                    $dtLayoutElements.show();
                   } else {
-                    dtable.rows().nodes().to$().hide();
-                    $(table).find("thead").hide();
-                    $(table).css("border-bottom", "unset");
-                    $(table).siblings(".dataTables_paginate").hide();
-                    $(table).siblings(".dataTables_info").hide();
+                    $dtLayoutElements.hide();
                   }
                 });

