diff --git a/core/modules/system/src/FileDownloadController.php b/core/modules/system/src/FileDownloadController.php
index 9bf889084a..c527aed07c 100644
--- a/core/modules/system/src/FileDownloadController.php
+++ b/core/modules/system/src/FileDownloadController.php
@@ -75,7 +75,7 @@ public function download(Request $request, $scheme = 'private') {
     // Merge remaining path arguments into relative file path.
     $uri = $scheme . '://' . $target;
 
-    if ($this->streamWrapperManager->isValidScheme($scheme) && file_exists($uri)) {
+    if ($this->streamWrapperManager->isValidScheme($scheme) && is_file($uri)) {
       // Let other modules provide headers and controls access to the file.
       $headers = $this->moduleHandler()->invokeAll('file_download', [$uri]);
 
