diff --git a/src/Plugin/Field/FieldWidget/EntityReferenceHierarchyAutocomplete.php b/src/Plugin/Field/FieldWidget/EntityReferenceHierarchyAutocomplete.php
index 39e5348116d1ccbcfdc5efb8c6e0f2b37416ec8e..ba2e58557074cb5dce8e3e0620a87c345b6c2ab9 100644
--- a/src/Plugin/Field/FieldWidget/EntityReferenceHierarchyAutocomplete.php
+++ b/src/Plugin/Field/FieldWidget/EntityReferenceHierarchyAutocomplete.php
@@ -72,7 +72,6 @@ class EntityReferenceHierarchyAutocomplete extends EntityReferenceAutocompleteWi
    */
   public function formElement(FieldItemListInterface $items, $delta, array $element, array &$form, FormStateInterface $form_state) {
     $widget = [
-      '#attributes' => ['class' => ['form--inline', 'clearfix']],
       '#theme_wrappers' => ['container'],
     ];
     $widget['target_id'] = parent::formElement($items, $delta, $element, $form, $form_state);
@@ -83,6 +82,7 @@ class EntityReferenceHierarchyAutocomplete extends EntityReferenceAutocompleteWi
       ];
     }
     else {
+      $widget['#attributes'] = ['class' => ['form--inline', 'clearfix']];
       $widget['weight'] = [
         '#type' => 'number',
         '#size' => '4',
diff --git a/src/Plugin/Field/FieldWidget/EntityReferenceHierarchySelect.php b/src/Plugin/Field/FieldWidget/EntityReferenceHierarchySelect.php
index 5ea15c88bb7c3a9124207202816e3d7b393174b8..8c055a31230f1617d39fca8f18dc34b0554123f7 100644
--- a/src/Plugin/Field/FieldWidget/EntityReferenceHierarchySelect.php
+++ b/src/Plugin/Field/FieldWidget/EntityReferenceHierarchySelect.php
@@ -66,7 +66,6 @@ class EntityReferenceHierarchySelect extends OptionsWidgetBase {
     ];
 
     $widget = [
-      '#attributes' => ['class' => ['form--inline', 'clearfix']],
       '#theme_wrappers' => ['container'],
     ];
 
@@ -78,6 +77,7 @@ class EntityReferenceHierarchySelect extends OptionsWidgetBase {
       ];
     }
     else {
+      $widget['#attributes'] = ['class' => ['form--inline', 'clearfix']];
       $widget['weight'] = [
         '#type' => 'number',
         '#size' => '4',
