diff --git a/modules/entity_hierarchy_breadcrumb/src/HierarchyBasedBreadcrumbBuilder.php b/modules/entity_hierarchy_breadcrumb/src/HierarchyBasedBreadcrumbBuilder.php
index d2a3739..97c3839 100644
--- a/modules/entity_hierarchy_breadcrumb/src/HierarchyBasedBreadcrumbBuilder.php
+++ b/modules/entity_hierarchy_breadcrumb/src/HierarchyBasedBreadcrumbBuilder.php
@@ -87,7 +87,7 @@ class HierarchyBasedBreadcrumbBuilder implements BreadcrumbBuilderInterface {
    * {@inheritdoc}
    */
   public function applies(RouteMatchInterface $route_match) {
-    if ($this->adminContext->isAdminRoute($route_match->getRouteObject())) {
+    if ($this->adminContext->isAdminRoute($route_match->getRouteObject()) && $route_match->getRouteName() !== 'entity.node.edit_form') {
       return FALSE;
     }
     $route_entity = $this->getEntityFromRouteMatch($route_match);
