diff --git a/config/schema/entity_reference_tree.schema.yml b/config/schema/entity_reference_tree.schema.yml
index f5df279..85e53f4 100644
--- a/config/schema/entity_reference_tree.schema.yml
+++ b/config/schema/entity_reference_tree.schema.yml
@@ -23,3 +23,15 @@ field.widget.settings.entity_reference_tree:
     dialog_title:
       type: string
       label: 'Dialog title'
+    auto_check_ancestors:
+      type: boolean
+      label: 'Automatically check ancestors when a node is checked'
+    auto_check_descendants:
+      type: boolean
+      label: 'Automatically check descendants when a node is checked'
+    auto_uncheck_ancestors:
+      type: boolean
+      label: 'Automatically uncheck ancestors when a node is unchecked'
+    auto_uncheck_descendants:
+      type: boolean
+      label: 'Automatically uncheck descendants when a node is unchecked'
diff --git a/js/entity_reference_tree.js b/js/entity_reference_tree.js
index fc917f0..ae7e9f6 100644
--- a/js/entity_reference_tree.js
+++ b/js/entity_reference_tree.js
@@ -7,7 +7,39 @@
 // @See https://www.drupal.org/docs/8/api/javascript-api/javascript-api-overview
 (function($, Drupal) {
   Drupal.behaviors.entityReferenceTree = {
+    applyAutoSelect: function (data, treeContainer, entityReferenceTreeWidgetConfig) {
+      const config = entityReferenceTreeWidgetConfig;
+      if (data.action == 'select_node') {
+        // Select ancestors when node is selected.
+        if (config.auto_check_ancestors) {
+          data.node.parents.forEach((parent) => {
+            treeContainer.jstree('select_node', parent);
+          });
+        }
+        // Select descendants when node is selected.
+        if (config.auto_check_descendants) {
+          data.node.children_d.forEach((child) => {
+            treeContainer.jstree('select_node', child);
+          });
+        }
+      }
+      if (data.action == 'deselect_node') {
+        // Deselect ancestors when node is deselected.
+        if (config.auto_uncheck_ancestors) {
+          data.node.parents.forEach((parent) => {
+            treeContainer.jstree('deselect_node', parent);
+          });
+        }
+        // Deselect descendants when node is deselected.
+        if (config.auto_uncheck_descendants) {
+          data.node.children_d.forEach((child) => {
+            treeContainer.jstree('deselect_node', child);
+          });
+        }
+      }
+    },
     attach: function(context, settings) {
+      const self = this;
       $("#entity-reference-tree-wrapper", context)
         .once("jstreeBehavior")
         .each(function() {
@@ -21,6 +53,11 @@
             const entityType = $("#entity-reference-tree-entity-type").val();
             const bundle = $("#entity-reference-tree-entity-bundle").val();
             const token = settings["entity_tree_token_" + fieldEditName];
+            const entityReferenceTreeWidgetConfig =
+              settings["entityReferenceTreeWidget"] ? (
+                settings["entityReferenceTreeWidget"][fieldEditName] ?
+                  settings["entityReferenceTreeWidget"][fieldEditName] : false
+              ) : false;
             const idIsString = bundle === "*";
             const limit = parseInt(settings["tree_limit_" + fieldEditName]);
             let selectedNodes;
@@ -97,7 +134,6 @@
               		// No limit.
               		return true;
               	}
-                
               },
               plugins: ["search", "changed", "checkbox", "conditionalselect"]
             });
@@ -112,7 +148,9 @@
               // selected node objects;
               const choosedNodes = data.selected;
               const r = [];
-
+              if (entityReferenceTreeWidgetConfig) {
+                self.applyAutoSelect(data, treeContainer, entityReferenceTreeWidgetConfig);
+              }
               for (let i = 0; i < choosedNodes.length; i++) {
                 const node = data.instance.get_node(choosedNodes[i]);
                 // node text escaping double quote.
diff --git a/src/Plugin/Field/FieldWidget/EntityReferenceTreeWidget.php b/src/Plugin/Field/FieldWidget/EntityReferenceTreeWidget.php
index af8511b..5f415e3 100644
--- a/src/Plugin/Field/FieldWidget/EntityReferenceTreeWidget.php
+++ b/src/Plugin/Field/FieldWidget/EntityReferenceTreeWidget.php
@@ -99,6 +99,12 @@ class EntityReferenceTreeWidget extends EntityReferenceAutocompleteWidget {
       ],
     ];
 
+    // Widget settings.
+    $settings = $this->getSettings();
+
+    // Pass data to js file.
+    $form['#attached']['drupalSettings']['entityReferenceTreeWidget'][$edit_id] = $settings;
+
     return $arr_element;
   }
 
@@ -122,6 +128,12 @@ class EntityReferenceTreeWidget extends EntityReferenceAutocompleteWidget {
         'label' => '',
         // Dialog title.
         'dialog_title' => '',
+        // Auto check.
+        'auto_check_ancestors' => FALSE,
+        'auto_check_descendants' => FALSE,
+        // Auto uncheck.
+        'auto_uncheck_ancestors' => FALSE,
+        'auto_uncheck_descendants' => FALSE,
     ] + parent::defaultSettings();
   }
 
@@ -168,6 +180,30 @@ class EntityReferenceTreeWidget extends EntityReferenceAutocompleteWidget {
       '#default_value' => $this->getSetting('dialog_title'),
     ];
 
+    $element['auto_check_ancestors'] = [
+      '#type' => 'checkbox',
+      '#title' => $this->t('Automatically check ancestors when a node is checked.'),
+      '#default_value' => $this->getSetting('auto_check_ancestors'),
+   ];
+
+    $element['auto_uncheck_ancestors'] = [
+      '#type' => 'checkbox',
+      '#title' => $this->t('Automatically uncheck ancestors when a node is unchecked.'),
+      '#default_value' => $this->getSetting('auto_uncheck_ancestors'),
+    ];
+
+    $element['auto_check_descendants'] = [
+      '#type' => 'checkbox',
+      '#title' => $this->t('Automatically check descendants when a node is checked.'),
+      '#default_value' => $this->getSetting('auto_check_descendants'),
+    ];
+
+    $element['auto_uncheck_descendants'] = [
+      '#type' => 'checkbox',
+      '#title' => $this->t('Automatically uncheck descendants when a node is unchecked.'),
+      '#default_value' => $this->getSetting('auto_uncheck_descendants'),
+    ];
+
     return $element;
   }
 
@@ -186,7 +222,27 @@ class EntityReferenceTreeWidget extends EntityReferenceAutocompleteWidget {
     if ($label = $this->getSetting('dialog_title')) {
       $summary[] = t('Dialog title: @title', ['@title' => $label]);
     }
-    
+
+    $summary[] = t('Check ancestors when node is checked: @yesno', [
+      '@yesno' => $this->getSetting('auto_check_ancestors') ?
+        $this->t('Yes') : $this->t('No')
+    ]);
+
+    $summary[] = t('Check descendants when node is checked: @yesno', [
+      '@yesno' => $this->getSetting('auto_check_descendants') ?
+        $this->t('Yes') : $this->t('No')
+    ]);
+
+    $summary[] = t('Uncheck ancestors when node is unchecked: @yesno', [
+      '@yesno' => $this->getSetting('auto_uncheck_ancestors') ?
+        $this->t('Yes') : $this->t('No')
+    ]);
+
+    $summary[] = t('Uncheck descendants when node is unchecked: @yesno', [
+      '@yesno' => $this->getSetting('auto_uncheck_descendants') ?
+        $this->t('Yes') : $this->t('No')
+    ]);
+
     return $summary;
   }
 }
