diff --git a/field_tokens.info.yml b/field_tokens.info.yml
index 79fc9e3..7bbd64b 100644
--- a/field_tokens.info.yml
+++ b/field_tokens.info.yml
@@ -1,6 +1,6 @@
 name: Field tokens
 description: Adds formatted field and field property tokens.
 type: module
-core: 8.x
+core_version_requirement: ^8 || ^9
 dependencies:
-  - token
+  - token:token
diff --git a/src/Tests/FieldTokensTestBase.php b/src/Tests/FieldTokensTestBase.php
index 57010ac..4d90a97 100644
--- a/src/Tests/FieldTokensTestBase.php
+++ b/src/Tests/FieldTokensTestBase.php
@@ -2,7 +2,7 @@
 
 namespace Drupal\field_tokens\Tests;
 
-use Drupal\image\Tests\ImageFieldTestBase;
+use Drupal\Tests\image\Functional\ImageFieldTestBase;
 
 /**
  * Class FieldTokensTest.
