diff --git a/lib/Drupal/Core/Session/AccessPolicyProcessor.php b/lib/Drupal/Core/Session/AccessPolicyProcessor.php
--- a/lib/Drupal/Core/Session/AccessPolicyProcessor.php
+++ b/lib/Drupal/Core/Session/AccessPolicyProcessor.php
@@ -46,14 +46,39 @@ public function processAccessPolicies(AccountInterface $account, string $scope =
 
     // If running in a fiber, prevent the current user switch from escaping to
     // outside the fiber by resuming the fiber if it was suspended.
-    $fiber = new \Fiber([$this, 'doProcessAccessPolicies']);
-    $fiber->start($account, $scope);
-    while (!$fiber->isTerminated()) {
-      if ($fiber->isSuspended()) {
-        $resume_type = $fiber->resume();
-        if (!$fiber->isTerminated() && $resume_type !== FiberResumeType::Immediate) {
-          usleep(500);
+    try {
+      $fiber = new \Fiber([$this, 'doProcessAccessPolicies']);
+      $fiber->start($account, $scope);
+      while (!$fiber->isTerminated()) {
+        if ($fiber->isSuspended()) {
+          $resume_type = $fiber->resume();
+          if (!$fiber->isTerminated() && $resume_type !== FiberResumeType::Immediate) {
+            usleep(500);
+          }
         }
       }
+      return $fiber->getReturn();
+    }
+    catch (\FiberError) {
+      // If fiber switching is disallowed, avoid account-switching and cache
+      // writes in this request path to prevent leaking switched users.
+      $calculated_permissions = new RefinableCalculatedPermissions();
+      foreach ($this->accessPolicies as $access_policy) {
+        if (!$access_policy->applies($scope)) {
+          continue;
+        }
+        $policy_permissions = $access_policy->calculatePermissions($account, $scope);
+        if (!$this->validateScope($scope, $policy_permissions)) {
+          throw new AccessPolicyScopeException(sprintf('The access policy "%s" returned permissions for scopes other than "%s".', get_class($access_policy), $scope));
+        }
+        $calculated_permissions = $calculated_permissions->merge($policy_permissions);
+      }
+      foreach ($this->accessPolicies as $access_policy) {
+        if ($access_policy->applies($scope)) {
+          $access_policy->alterPermissions($account, $scope, $calculated_permissions);
+        }
+      }
+      $calculated_permissions->addCacheTags(['access_policies']);
+      return new CalculatedPermissions($calculated_permissions);
     }
 
-    return $fiber->getReturn();
   }
