diff --git a/type_tray.module b/type_tray.module
--- a/type_tray.module
+++ b/type_tray.module
@@ -125,7 +125,7 @@ function type_tray_theme($existing, $type, $theme, $path) {
         'items' => NULL,
         'layout' => NULL,
         'category_labels' => NULL,
-        'node_add_page_url' => Url::fromRoute('node.add_page')->toString(),
+        'node_add_page_url' => Url::fromRoute('entity.node.add_page')->toString(),
       ],
     ],
   ];
