diff --git a/js/csv-field.js b/js/csv-field.js
index 30ecbe8..4b44990 100644
--- a/js/csv-field.js
+++ b/js/csv-field.js
@@ -298,7 +298,7 @@
             tableSettings.columns.push({ title: headerData[i] });
           }
 
-          tableSettings.autoWidth = true;
+          tableSettings.autoWidth = false;
 
           // If "First column of table is a row header" is checked, tag the
           // first column cells as <th> elements instead of <td>.
