--- a/src/Plugin/Field/FieldType/EntityReferenceRevisionsItem.php
+++ b/src/Plugin/Field/FieldType/EntityReferenceRevisionsItem.php
@@ -212,8 +212,11 @@
    */
   public function getValue() {
     $values = parent::getValue();
-    if ($this->entity instanceof EntityNeedsSaveInterface && $this->entity->needsSave()) {
-      $values['entity'] = $this->entity;
+    if ($this->isEntityLoaded()) {
+      $entity = $this->properties['entity']->getValue();
+      if ($entity instanceof EntityNeedsSaveInterface && $entity->needsSave()) {
+        $values['entity'] = $entity;
+      }
     }
     return $values;
   }
