diff --git a/src/ConnectionFactory.php b/src/ConnectionFactory.php
index 41b21b1d1ca9c0a1bc5c622d967845ac4aef7012..0a45a6ea674956be86dd69fb5f37621749ccc58e 100644
--- a/src/ConnectionFactory.php
+++ b/src/ConnectionFactory.php
@@ -82,6 +82,11 @@ class ConnectionFactory {
         $options['driverOptions'] = $info['pdo'];
       }
       $this->cache[$target] = DriverManager::getConnection($options, new Configuration());
+      if (isset($info['init_commands'])) {
+        foreach ($info['init_commands'] as $sql) {
+          $this->cache[$target]->executeStatement($sql);
+        }
+      }
       if ($driver == 'sqlite') {
         $this->sqliteDatabases($this->cache[$target], $info['prefix'], $info['database']);
       }
