diff --git a/src/Plugin/video_embed_field/Provider/YouTube.php b/src/Plugin/video_embed_field/Provider/YouTube.php
index 224edd2e2d6f0157b77968877489ae05f953101d..ecbc2a8426b07f3056845817c4779da7b88330bc 100644
--- a/src/Plugin/video_embed_field/Provider/YouTube.php
+++ b/src/Plugin/video_embed_field/Provider/YouTube.php
@@ -87,7 +87,7 @@ class YouTube extends ProviderPluginBase {
    * {@inheritdoc}
    */
   public static function getIdFromInput($input) {
-    preg_match('/^https?:\/\/(www\.)?((?!.*list=)youtube\.com\/watch\?.*v=|youtu\.be\/)(?<id>[0-9A-Za-z_-]*)/', $input, $matches);
+    preg_match('/^https?:\/\/(www\.)?((?!.*list=)youtube\.com\/(watch(\?.*v=|\/)|shorts\/)|youtu\.be\/)(?<id>[0-9A-Za-z_-]*)/', $input, $matches);
     return isset($matches['id']) ? $matches['id'] : FALSE;
   }
 
diff --git a/tests/src/Kernel/MockHttpClient.php b/tests/src/Kernel/MockHttpClient.php
index d555d1f6a6a555262221a9ee25cb523d4f82d33f..30079d11c0799cb599343b466302bde76b03d8da 100644
--- a/tests/src/Kernel/MockHttpClient.php
+++ b/tests/src/Kernel/MockHttpClient.php
@@ -47,6 +47,8 @@ class MockHttpClient implements ClientInterface {
 
   /**
    * {@inheritdoc}
+   *
+   * @return mixed
    */
   public function getConfig(?string $option = NULL) {
     throw new \Exception(static::EXCEPTION_MESSAGE);
diff --git a/tests/src/Unit/ProviderUrlParseTest.php b/tests/src/Unit/ProviderUrlParseTest.php
index 6f64b970c1c95d500dc554252ab84e41e8ea6138..3bf0656fddc1a9763b7fbd1cf13b194f5256318f 100644
--- a/tests/src/Unit/ProviderUrlParseTest.php
+++ b/tests/src/Unit/ProviderUrlParseTest.php
@@ -57,6 +57,11 @@ class ProviderUrlParseTest extends UnitTestCase {
         'https://youtu.be/fdbFVWupSsw',
         'fdbFVWupSsw',
       ],
+      'YouTube: Short Full URL' => [
+        'Drupal\video_embed_field\Plugin\video_embed_field\Provider\YouTube',
+        'https://youtube.com/shorts/fdbFVWupSsw',
+        'fdbFVWupSsw',
+      ],
       'YouTube: With Language Preference' => [
         'Drupal\video_embed_field\Plugin\video_embed_field\Provider\YouTube',
         'https://youtube.com/watch?v=fdbFV_Wup-Ssw&hl=fr-ca',
